suite/tests/client-interface/memory_dump_test.dll.c

```diff
@@ -0,0 +1,69 @@
+/* **********************************************************
+ * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "dr_api.h"
+#include "client_tools.h"
+
+#include "dr_api.h"
+#include "client_tools.h"
+
+static bool saw_thread_init_event = false;
+
+static void
+dr_exit(void)
+{
+    if (!saw_thread_init_event)
+        dr_fprintf(STDERR, "Error: never saw thread init event!\n");
+}
+
+static void
+dr_thread_init(void *drcontext)
+{
+    dr_fprintf(STDERR, "thread init\n");
+    saw_thread_init_event = true;
+
+    dr_memory_dump_spec_t spec;
+    spec.size = sizeof(dr_memory_dump_spec_t);
+    spec.flags = DR_MEMORY_DUMP_ELF;
+    if (!dr_create_memory_dump(&spec))
```

Comment:
Hmm, this is not allowed according to:

 * This routine may not be called from any registered event callback
 * other than the nudge event or the pre- or post-system call event.
 * It may be called from clean calls out of the cache.
 * This routine may not be called while any locks are held that
 * could block a thread processing a registered event callback or cache
 * callout.

I think the callback_registration_lock is released before calling: so no locks are held?  Or is there one in the thread init caller?  Debug build should assert if locks are held: but see below: this test isn't even built or run it looks like.

There has to be a valid mcontext I think: is that not there for the 1st thread's init event?  #1152?  Though the suspend passes THREAD_SYNCH_NO_LOCKS_NO_XFER so actually no mcontext is needed?

Need to either figure out why it's not allowed in order to show it's safe and update the docs, or need to change the test.

Reply:
The debug build detects a lock order issue in using memquery_iterator_start()  (and related functions) and the string hash table. I've removed the string hash table, and use strstr() to address the issue.

PTAL.

Reply:
Your comment seems unrelated to mine.  Mine is about the state when the client calls dr_create_memory_dump.  Your comment is about internal activity inside the implementation.  For your comment, if it truly is single-threaded you could use a hashtable with no lock or you could use a different  (new) lock with a different rank.

Reply:
My comment was not addressed: I believe you are accepting the state limitations of when this can be called, but you're violating that here still by calling it in the thread init event.

Reply:
I've updated the code to use nudge event.